File: java/src/main/java/com/example/employeeschedulemanagementsys/model/Employee.java
package com.example.employeeschedulemanagementsys.model;

import java.util.*;

public class Employee {

    private final String name;
    private final Map<Day, List<Shift>> preferences = new EnumMap<>(Day.class);
    private boolean[][] assignments = new boolean[7][3];
    private int daysWorked = 0;

    public Employee(String name) {
        this.name = name;
    }

    public String getName() { return name; }
    public int getDaysWorked() { return daysWorked; }


    /** Store ranked preferences for one day. Nulls are silently ignored. */
    public void setPreferences(Day day, Shift first, Shift second, Shift third) {
        List<Shift> ranked = new ArrayList<>();
        if (first  != null) ranked.add(first);
        if (second != null) ranked.add(second);
        if (third  != null) ranked.add(third);
        preferences.put(day, ranked);
    }

    /** Preferences in priority order (index 0 = highest). Empty list if none set. */
    public List<Shift> getPreferences(Day day) {
        return preferences.getOrDefault(day, Collections.emptyList());
    }

    public boolean hasPreference(Day day) {
        return !getPreferences(day).isEmpty();
    }


    /**
     * True if the employee has not yet been assigned on this day AND
     * has not reached the 5-day weekly cap.
     */
    public boolean isAvailable(Day day) {
        if (daysWorked >= 5) return false;
        for (Shift s : Shift.values()) {
            if (assignments[day.ordinal()][s.ordinal()]) return false;
        }
        return true;
    }


    public void assign(Day day, Shift shift) {
        assignments[day.ordinal()][shift.ordinal()] = true;
        daysWorked++;
    }

    public void reset() {
        assignments = new boolean[7][3];
        daysWorked = 0;
    }

    @Override public String toString() { return name; }
}

File: jsapp/model/Employee.js

import DAYS   from "./Day.js";
import SHIFTS from "./Shift.js";

export default class Employee {

    constructor(name) {
        this.name        = name;
        this.preferences = {};                              // Map<day, Shift[]>
        this.assignments = Array.from(                      // boolean[7][3]
            { length: 7 },
            () => [false, false, false]
        );
        this.daysWorked  = 0;
    }

     /**
     * mirrors: setPreferences(Day day, Shift first, Shift second, Shift third)
     * Null/undefined values are filtered out — same as Java null-guard.
     */
    setPreferences(day, first, second, third) {
        const ranked = [first, second, third].filter(s => s != null);
        this.preferences[day] = ranked;
    }

    /**
     * mirrors: getPreferences(Day day) → List<Shift>
     * Returns preferences in priority order. Empty array if none set.
     */
    getPreferences(day) {
        return this.preferences[day] ?? [];
    }

    /**
     * mirrors: hasPreference(Day day)
     */
    hasPreference(day) {
        return this.getPreferences(day).length > 0;
    }

    // ── Availability ───────────────────────────────────────────────────

    /**
     * mirrors: isAvailable(Day day)
     * True if under 5-day cap AND not already assigned on this day.
     */
    isAvailable(day) {
        if (this.daysWorked >= 5) return false;           // 5-day cap
        const di = DAYS.indexOf(day);
        for (let si = 0; si < SHIFTS.length; si++) {
            if (this.assignments[di][si]) return false;     // already working today
        }
        return true;
    }

    // ── Assignment ─────────────────────────────────────────────────────

    /**
     * mirrors: assign(Day day, Shift shift)
     * Uses indexOf() in place of Java's .ordinal()
     */
    assign(day, shift) {
        this.assignments[DAYS.indexOf(day)][SHIFTS.indexOf(shift)] = true;
        this.daysWorked++;
    }

    /**
     * mirrors: reset()
     * Called at the start of each buildSchedule() run.
     */
    reset() {
        this.assignments = Array.from(
            { length: 7 },
            () => [false, false, false]
        );
        this.daysWorked = 0;
    }
}